src/mpi/init/init_thread_cs.c

```diff
@@ -28,24 +28,26 @@ void MPII_init_thread_and_enter_cs(int thread_required)
     MPID_Thread_mutex_create(&MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX, &err);
     MPIR_Assert(err == 0);
 
-    /* We don't know requested or provided thread level yet, set isThreaded
-     * to TRUE so we'll protect the init with critical section anyway.
-     */
-    /* To ensure consistency, we save the required isThreaded */
-    required_is_threaded = (thread_required == MPI_THREAD_MULTIPLE);
-    MPIR_ThreadInfo.isThreaded = required_is_threaded;
+    MPIR_ThreadInfo.isThreaded = (thread_required == MPI_THREAD_MULTIPLE);
+    required_is_threaded = MPIR_ThreadInfo.isThreaded;
+
     MPID_THREAD_CS_ENTER(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
     MPID_THREAD_CS_ENTER(VCI, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
+    /* Reset isThreaded during init to disable all other mutexes.
+     * One, they may not have been initialized.
+     * Two, there is no need for additional mutex locks during init.
+     */
```

Comment:
I am not sure if this logic is really correct. In my understanding, if we set `MPIR_ThreadInfo.isThreaded = 0;` here, any other threads that try to take `MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX` will ignore a mutex in `MPID_THREAD_CS_ENTER`. Is this behavior okay?

Reply:
The entire init code is already locked under `MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX`. MPI standards forbids most MPI calls before `MPI_Init` or `MPI_Finalize`, other than trivial ones such as `MPI_Initialized` and `MPI_Finalized`.

Reply:
In another word, if user tried to run `MPI_Send` in another thread before `MPI_init` finish, it is a user error.

Reply:
I see, then do we need to take a lock here (MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX)?
Because  `MPIR_ThreadInfo.isThreaded` is set to 0, this lock does not protect anything.

Reply:
On second thought, you are right. It essentially voids the lock in `MPI_Init` itself. I'll work out another strategy.